src/spotless.ts

```diff
@@ -1,54 +1,94 @@
+];
 
 function sanitizePath(fsPath: string): string {
-  const platform = process.platform;
-  if (platform === 'win32') {
-    // vscode.Uri.fsPath will lowercase the drive letters
+  if (process.platform === 'win32') {
+    // vscode.Uri.fsPath will lower-case the drive letters
     // https://github.com/microsoft/vscode/blob/dc348340fd1a6c583cb63a1e7e6b4fd657e01e01/src/vs/vscode.d.ts#L1338
     return fsPath[0].toUpperCase() + fsPath.substr(1);
   }
   return fsPath;
 }
 
+interface TextBuffers {
+  stdOut: string[];
+  stdErr: string[];
+}
+
 export async function makeSpotless(
   gradleApi: GradleTasksApi,
   document: vscode.TextDocument
-): Promise<null> {
+): Promise<string | null> {
   if (document.isClosed || document.isUntitled) {
-    logger.warning('Document is closed or not saved yet, skipping');
-    return null;
+    throw new Error('Document is closed or not saved yet, skipping');
   }
   const workspaceFolder = vscode.workspace.getWorkspaceFolder(document.uri);
   if (!workspaceFolder) {
-    logger.error(
+    throw new Error(
       `Unable to find workspace folder for ${path.basename(
         document.uri.fsPath
       )}, skipping`
     );
-    return null;
   }
-  logger.info('Running spotless on', path.basename(document.uri.fsPath));
-  try {
-    if (document.isDirty) {
-      const saved = await document.save();
-      if (!saved) {
-        throw new Error(`Unable to save file ${document.fileName}`);
+
+  const sanitizedPath = sanitizePath(document.uri.fsPath);
+  const args = [`-PspotlessIdeHook=${sanitizedPath}`, '--quiet'];
+  const textBuffers: TextBuffers = {
+    stdOut: [],
+    stdErr: [],
+  };
+
+  const runTaskOpts: RunTaskOpts = {
+    projectFolder: workspaceFolder.uri.fsPath,
+    taskName: 'spotlessApply',
+    args,
+    showProgress: true,
+    input: document.getText(),
+    showOutputColors: false,
+    onOutput: (output: Output) => {
+      switch (output.getOutputType()) {
+        case Output.OutputType.STDOUT:
+          textBuffers.stdOut.push(output.getMessage());
+          break;
+        case Output.OutputType.STDERR:
+          textBuffers.stdErr.push(output.getMessage());
+          break;
       }
-    }
-    const normalizedPath = escapeStringRegexp(
-      sanitizePath(document.uri.fsPath)
-    );
-    const args = [`-PspotlessFiles=${normalizedPath}`];
-    await gradleApi.runTask(workspaceFolder.uri.fsPath, 'spotlessApply', args);
-  } catch (e) {
-    logger.error('Error running spotless formatter:', e.message);
+    },
+  };
+
+  await gradleApi.runTask(runTaskOpts);
+
+  const stdOut = textBuffers.stdOut.join('');
+  const stdErr = textBuffers.stdErr.join('').trim();
+
+  if (SPOTLESS_STATUSES.includes(stdErr)) {
+    logger.debug('Spotless:', stdErr);
+  }
+  if (stdErr === SPOTLESS_STATUS_IS_DIRTY) {
+    return stdOut;
+  }
+  if (
+    stdErr !== SPOTLESS_STATUS_DID_NOT_CONVERGE &&
+    stdErr !== SPOTLESS_STATUS_IS_CLEAN
+  ) {
+    logger.error(stdErr);
```

Comment:
throw here

Reply:
If the status is `DID NOT CONVERGE`, you probably want to tell the user about it.  Usually it will happen on just one file, so it will seem like a bug that other files are being formatted, but just this one is not.  Probably everything besides `STATUS_IS_CLEAN/IS_DIRTY` ought to be reported to the user in some way, and a raw dump of `stderr` would be okay and informative.

Reply:
Thanks for the review. I agree. I'll log everything that isn't `STATUS_IS_CLEAN/IS_DIRTY`. Curious though, what exactly does `DID NOT CONVERGE` mean? Just thinking about how best to present this state to the user.

Reply:
Example:

```
run 1: public class A {}
run 2: public  class  A  {}
run 3: public   class   A   {}
run 4: public    class    A    {}
```

Spotless always checks for this condition, and can silently resolve some of them.  We always design our error messages to be user-presentable:

https://github.com/diffplug/spotless/blob/13700cbe45344f45cbef79cbba94aec6724757f8/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/IdeHook.java#L51-L52

The document we point users to:

https://github.com/diffplug/spotless/blob/master/PADDEDCELL.md#a-misbehaving-step

Reply:
I've decided to keep it simple and just log the spotless statuses:

<img width="764" alt="Screenshot 2020-05-10 at 14 27 31" src="https://user-images.githubusercontent.com/102141/81500549-be2eab00-92ca-11ea-8ced-bedf5aef9b45.png">

If status is `DID NOT CONVERGE`, it will be shown in the same output pannel as:
`[error] App.java DID NOT CONVERGE`

I've not been able to replicate this scenario which is why I can't provide a screenshot for it. 

I'll add some info about this state to the README and also link out to the PADDEDCELL docs.

Reply:
Actually no need to add this info to the docs as the error message will already contain this info: https://github.com/diffplug/spotless/blob/13700cbe45344f45cbef79cbba94aec6724757f8/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/IdeHook.java#L52